### ISSUE TITLE
Updates to make

### ISSUE THREAD

**Issue by michalkrupa** [michalkrupa/kln#3]
- Expand on all of the sections, esp. intro and related work and discussion to contextualize it a lot more. for instance the intro parts don't set up the problem statement enough, when talking about structure dependency give some clear formalism in general notation that explains what it is you're trying to recover, why is this problem something other researchers or engineers care about? 

- explain more what are the traditional approaches and expand on these in related work, how do they work? what sort of failure modes do they have (either failure with respect to performance or understanding)? and why that motivates the method

- explain further how KL divergence frameworks play into it, or what you're doing with them, and if you're training using MSE and not KL then it's not clear what KL is doing here

- methods section should have more details and notation to make clear what is happening i.e. when introducing things like simple MLP that still needs to be made clear what it is (deep? Shallow? Activation function choices?)

- when introducing notation like i, j, k are they scalars? Vector embeddings symbolically represented by these variables? give the dimensionality of them (in latex this is done by \mathbb{R} to indicate the reals and then the exponent put on that indicate dimensionality, like k \in \mathbb{R} means learning a scalar target but k \in \mathbb{R}^c means you learning a c dimensional output

Inferred dealing with scalars for 4.1 and 4.3 but then not sure if it's the case in 4.4 since those embeddings are probably some d dim embeddings? And not clear what the targets are in 4.4

- How is train/test split defined; is optimization determined based on a validation set and then  final report held out test results?

- What loss is optimized in each experiment?

- Figure 1 caption says learned mapping of k but also says true mapping of k? Doesn't seem like it can be both at the same time on the one plot

General takeaways:
#### The ML approach:
- trying to regress from two variables (or two vectors) i, j onto a scalar target. want to blend a linear model and a non linear model which are jointly learned by an alpha factor which interpolates the two. set up two synthetic datasets and one real world dataset to explore this setting and find that the model puts about equal weight on both models in setup but doesn't quite reach the performance of the MLP on its own.

- From an experimental design POV it looks like  reporting only train loss, but need to  eventually start thinking about train/val/test splits in order to generalize processes and be more rigorous in any reported accuracy or loss. For instance, it's not unreasonable to expect optimizing KLN for the same number of epochs on the same optimization schedule is a little worse on train loss than just an MLP because in the former you're optimizing 2x the number of parameters. Maybe need to check what happens when comparing KLN after more epochs or you need to try other optimization schedules for KLN. Or separately train MLP and KLN on data until it reaches the same train loss (or halt MLP earlier when it's at the lowest train loss achievable by KLN) and then test both on a held out test sample, then ask: which method performs better conditioned on achieving the same train loss. It's a different question than current plots are indicating but could be illuminating to try and understand the thing better. Then the second order question might be: how to train the KLN to the same train loss as an MLP?

**Comment by michalkrupa** [comment]
KL was used in the pathfinder algorithm I was trying to expand upon for curvature path constraints. I used KL there to reduce to n (10) nearest neighbors because the graph traversal was taking forever otherwise, even with djikstras. Will work through these concerns before publishing the next draft
